### PR TITLE
implement new queries and change donation creation flow

### PR DESCRIPTION
**Pull request by mohammadranjbarz**
related to Giveth/giveth-dapps-v2#630

### CODE REVIEW

**Comment by mohammadranjbarz** [comment]
After merging this PR the frontend should stop using `savedonation` mutation and instead of that it should call 
`createDonation` and the sent data is less than `saveDonation`
* you just have to sent these data
```
 mutation (
    $transactionId: String!
    $transactionNetworkId: Float!
    $nonce: Float!
    $amount: Float!
    $token: String!
    $projectId: Float!
    $transakId: String
    $tokenAddress: String
    $anonymous: Boolean
  ) {
    createDonation(
      transactionId: $transactionId
      transactionNetworkId: $transactionNetworkId
      nonce: $nonce
      amount: $amount
      token: $token
      projectId: $projectId
      transakId: $transakId
      tokenAddress: $tokenAddress
      anonymous: $anonymous
    )
  }
```
*  you don't need to send `fromWalletAddress` we extract it from users's token
* user can't call it if it's not logged in, and it should be logged-in with wallet address that want to donate
* you don't need to send `toWalletAddress` we get it from projects info ( you send projected as you see)
* **We don't send email to project owners and donors in this state, when donations got verified by network the we send the emails and segment events**

After frontend notified that transaction has been minded it should call this query

```
  query (
    $donationId: Float!
  ) {
    inquiryDonationWithNetwork(
      donationId: $donationId
    ){
      id
      status
      verifyErrorMessage
    
    }
  }
```
when calling it we check the donation with network transaction and we return the updated donation in response ( as I said above, in this state we send emails to donors and project owners, and if user didn't call this webservice we have a job that runs every hour and verify donations with network, ...)

And if frontend knew that tx has been failed it should call this:
```
  mutation (
    $status: String!
    $donationId: Float!
  ) {
    updateDonationStatus(
      status: $status
      donationId: $donationId
    ){
      id
      status
      verifyErrorMessage
    }
  }
```
and send `failed` as status, in this case we heck donation with network and if it's still `pending` we trust the user and change the donation status to `failed`

**We would delete saveDonation mutation when frontend changed to use these new srvices**

**PS** user should be logged-in for all these queries/mutations and loggedIn userId should be the donations.userId

**Comment by mohammadranjbarz** [comment]
> After merging this PR the frontend should stop using `savedonation` mutation and instead of that it should call `createDonation` and the sent data is less than `saveDonation`
> 
> * you just have to sent these data
> 
> ```
>  mutation (
>     $transactionId: String!
>     $transactionNetworkId: Float!
>     $nonce: Float!
>     $amount: Float!
>     $token: String!
>     $projectId: Float!
>     $transakId: String
>     $tokenAddress: String
>     $anonymous: Boolean
>   ) {
>     createDonation(
>       transactionId: $transactionId
>       transactionNetworkId: $transactionNetworkId
>       nonce: $nonce
>       amount: $amount
>       token: $token
>       projectId: $projectId
>       transakId: $transakId
>       tokenAddress: $tokenAddress
>       anonymous: $anonymous
>     )
>   }
> ```
> 
> * you don't need to send `fromWalletAddress` we extract it from users's token
> * user can't call it if it's not logged in, and it should be logged-in with wallet address that want to donate
> * you don't need to send `toWalletAddress` we get it from projects info ( you send projected as you see)
> * We don't send email to project owners and donors in this state, when donations got verified by network the we send the emails and segment events
> 
> After frontend notified that transaction has been minded it should call this query
> 
> ```
>   query (
>     $donationId: Float!
>   ) {
>     inquiryDonationWithNetwork(
>       donationId: $donationId
>     ){
>       id
>       status
>       verifyErrorMessage
>     
>     }
>   }
> ```
> 
> when calling it we check the donation with network transaction and we return the updated donation in response ( as I said above, in this state we send emails to donors and project owners, and if user didn't call this webservice we have a job that runs every hour and verify donations with network, ...)
> 
> And if frontend knew that tx has been failed it should call this:
> 
> ```
>   mutation (
>     $status: String!
>     $donationId: Float!
>   ) {
>     updateDonationStatus(
>       status: $status
>       donationId: $donationId
>     ){
>       id
>       status
>       verifyErrorMessage
>     }
>   }
> ```
> 
> and send `failed` as status, in this case we heck donation with network and if it's still `pending` we trust the user and change the donation status to `failed`
> 
> **PS** user should be logged-in for all these queries/mutations and loggedIn userId should be the donations.userId

@MoeNick  @RamRamez 
plz take a look on what I did

**Comment by MoeNick** [comment]
I love it @mohammadranjbarz, what about updating the trx?

**Comment by mohammadranjbarz** [comment]
@mateodaza See here please

**Comment by mateodaza** [comment]
Thanks @mohammadranjbarz this looks good to me. Would love @RamRamez input also

**Comment by CarlosQ96** [comment]
> After merging this PR the frontend should stop using `savedonation` mutation and instead of that it should call `createDonation` and the sent data is less than `saveDonation`
> 
> * you just have to sent these data
> 
> ```
>  mutation (
>     $transactionId: String!
>     $transactionNetworkId: Float!
>     $nonce: Float!
>     $amount: Float!
>     $token: String!
>     $projectId: Float!
>     $transakId: String
>     $tokenAddress: String
>     $anonymous: Boolean
>   ) {
>     createDonation(
>       transactionId: $transactionId
>       transactionNetworkId: $transactionNetworkId
>       nonce: $nonce
>       amount: $amount
>       token: $token
>       projectId: $projectId
>       transakId: $transakId
>       tokenAddress: $tokenAddress
>       anonymous: $anonymous
>     )
>   }
> ```
> 
> * you don't need to send `fromWalletAddress` we extract it from users's token
> * user can't call it if it's not logged in, and it should be logged-in with wallet address that want to donate
> * you don't need to send `toWalletAddress` we get it from projects info ( you send projected as you see)
> * **We don't send email to project owners and donors in this state, when donations got verified by network the we send the emails and segment events**
> 
> After frontend notified that transaction has been minded it should call this query
> 
> ```
>   query (
>     $donationId: Float!
>   ) {
>     inquiryDonationWithNetwork(
>       donationId: $donationId
>     ){
>       id
>       status
>       verifyErrorMessage
>     
>     }
>   }
> ```
> 
> when calling it we check the donation with network transaction and we return the updated donation in response ( as I said above, in this state we send emails to donors and project owners, and if user didn't call this webservice we have a job that runs every hour and verify donations with network, ...)
> 
> And if frontend knew that tx has been failed it should call this:
> 
> ```
>   mutation (
>     $status: String!
>     $donationId: Float!
>   ) {
>     updateDonationStatus(
>       status: $status
>       donationId: $donationId
>     ){
>       id
>       status
>       verifyErrorMessage
>     }
>   }
> ```
> 
> and send `failed` as status, in this case we heck donation with network and if it's still `pending` we trust the user and change the donation status to `failed`
> 
> **We would delete saveDonation mutation when frontend changed to use these new srvices**
> 
> **PS** user should be logged-in for all these queries/mutations and loggedIn userId should be the donations.userId

Totally Agree with this changes!

File: test/testUtils.ts [hunk 1]
```diff
@@ -2,7 +2,7 @@ import { assert } from 'chai';
 import * as jwt from 'jsonwebtoken';
 import config from '../src/config';
 import { NETWORK_IDS } from '../src/provider';
-import { User } from '../src/entities/user';
+import { User, UserRole } from '../src/entities/user';
```
Comment (CarlosQ96) [inline test/testUtils.ts:5]: UserRole is unused 😄

**Review by CarlosQ96 — APPROVED** [review]
I checked the code and tests, and agree with the refactors, it's a lot! Great job!
Much more robust saveDonation flow

**Comment by RamRamez** [comment]
Thanks @mohammadranjbarz for great work.
I have a suggestion, can we merge `inquiryDonationWithNetwork` and `updateDonationStatus`?
When tx is mined or failed, both are kind of updating  donation status and we should be able to use `updateDonationStatus` for both, so calling `inquiryDonationWithNetwork` is unnecessary.

@mohammadranjbarz @CarlosQ96

**Comment by mohammadranjbarz** [comment]
> Thanks @mohammadranjbarz for great work. I have a suggestion, can we merge `inquiryDonationWithNetwork` and `updateDonationStatus`? When tx is mined or failed, both are kind of updating donation status and we should be able to use `updateDonationStatus` for both, so calling `inquiryDonationWithNetwork` is unnecessary.
> 
> @mohammadranjbarz @CarlosQ96

@RamRamez  Thank for giving feedback, yeah we can merge to one web service, to be honest already these two webservices have lots of repetitive codes, so it would make our backend code cleaner, so You just need to call `updateDonationStatus` and the `status` would be optional and it's not necessary anymore ( if you don't send status, we just check it with network and update it if it's necessary)

**Comment by mohammadranjbarz** [comment]
@RamRamez Now we merged this PR, so you can integrate to frontend with it

**Comment by RamRamez** [comment]
@mohammadranjbarz 
What if front-end tells a lie to backend about donation status? so backend should check status with network anyway.
I have suggestion, we can remove `status` input in `updateDonationStatus`.
If donation is failed or mined, front-end just triggers `updateDonationStatus` with `donationId` to inform backend to update donation status with network.

**Comment by mohammadranjbarz** [comment]
> @mohammadranjbarz What if front-end tells a lie to backend about donation status? so backend should check status with network anyway. I have suggestion, we can remove `status` input in `updateDonationStatus`. If donation is failed or mined, front-end just triggers `updateDonationStatus` with `donationId` to inform backend to update donation status with network.

 We can transaction info from network when it's mined, so if a transaction is not mined or rejected we can't know ( maybe we can check with nonce and if donor has initiated another transaction after that  we know that it's failed)
 
 So in the rejection and failed id front doesn't tell the backend that it's rejected, backend would not know currently, so it's better to notify backend in this case

**Pull request merged**